refactor: split cloud providers into separate modules

philip-gai opened the pull request:
## Summary

This PR refactors the Terraform configuration by splitting each cloud provider (AWS, Azure, GCP) into its own module under `src/`. This improves maintainability, allows for independent versioning, and makes it easier to use only the cloud providers you need.

## Changes

### Structure
- Created `src/aws/` module for AWS OIDC configuration
- Created `src/azure/` module for Azure OIDC configuration  
- Created `src/gcp/` module for GCP OIDC configuration
- Each module is self-contained with its own `main.tf`, `variables.tf`, `outputs.tf`, `provider.tf`, and `versions.tf`

### Code Quality
- AWS: Created `aws_client_id_list` local to avoid duplicating client IDs
- Azure: Removed unnecessary local variables, inlined values used only once
- GCP: Fixed variable description and used lowercase region default

### Dependabot
- Updated `.github/dependabot.yml` to track provider updates in each module directory

## Testing
- All modules pass `terraform fmt -check`
- All modules pass `terraform validate`

philip-gai committed:
Split out modules for each cloud provider

philip-gai commented on .devcontainer/devcontainer.json:
```diff
@@ -3,8 +3,7 @@
         "ghcr.io/devcontainers/features/terraform:1": {},
         "ghcr.io/devcontainers/features/aws-cli:1": {},
-        "ghcr.io/devcontainers/features/azure-cli:1": {},
-        "ghcr.io/bartventer/arch-devcontainer-features/gcloud-cli:1": {}
```
Remove this feature because it broke the image build for some reason

philip-gai committed:
Make thumbprint lowercase

philip-gai committed:
Add default sts.amazonaws.com" to trust relationship

philip-gai committed:
Add back examples

philip-gai committed:
Add default sts.amazonaws.com to provider

philip-gai committed:
Remove extra line

philip-gai committed:
Set default AWS_STS_ENDPOINT

lucavallin commented on src/aws/main.tf:
```diff
@@ -9,7 +28,7 @@ resource "aws_s3_bucket" "this" {
 resource "aws_iam_openid_connect_provider" "this" {
   url             = local.oidc_issuer_uri
-  client_id_list  = [local.aws_sts_endpoint]
+  client_id_list  = ["sts.amazonaws.com", local.aws_sts_endpoint]
```
Should `sts.amazonaws.com` always be in the client list?

philip-gai replied:
Yes, `sts.amazonaws.com` is now always included. I've created a local variable `aws_client_id_list` that includes both `sts.amazonaws.com` and the regional STS endpoint, and this is used in both the OIDC provider and the IAM role policy.

lucavallin commented on src/aws/main.tf:
```diff
@@ -28,7 +47,11 @@ resource "aws_iam_role" "this" {
-            "${aws_iam_openid_connect_provider.this.url}:aud" = local.aws_sts_endpoint
+            "${aws_iam_openid_connect_provider.this.url}:sub" : "${local.ghes_hostname}",
+            "${aws_iam_openid_connect_provider.this.url}:aud" = [
+              "sts.amazonaws.com",
```
We should create a `local` var for this

philip-gai replied:
Done! Created `aws_client_id_list` local variable that's now used in both the `client_id_list` attribute and the IAM role policy `aud` condition.

lucavallin commented on src/aws/provider.tf:
```diff
@@ -1,14 +1,6 @@
-provider "azurerm" {
```
Do we need to update the dependabot config to keep it updating the providers?

lucavallin commented on src/azure/main.tf:
```diff
@@ -1,3 +1,24 @@
+  azure_blob_endpoint_suffix             = "core.windows.net"
+}
+
+resource "random_string" "long" {
```
Since we use this across modules, we could have modules take in a random prefix as a variable, and generate the random prefix in our main.tf

lucavallin reviewed (commented):
Thank you for the contribution! This is a much needed upgrade. I left a few comments. Also, modules normally follow a certain structure (e.g. modules are inside a "modules" directory outside of the main "src" directory). Have a look at https://www.env0.com/blog/terraform-modules

bschaatsbergen reviewed (commented):
Hey 👋🏼, nice job on splitting this out. I left a few drive-by comments. They’re just meant to highlight some things for consideration.

philip-gai committed:
Address PR review comments

- AWS: Create local var aws_client_id_list to avoid duplication
- Azure: Remove unnecessary local vars (oidc_issuer_uri, azure_blob_endpoint_suffix)
- GCP: Fix variable description and use lowercase region default
- Update dependabot config to include module directories

philip-gai commented:
Thanks for the feedback! I've addressed all the review comments in the latest commit:

**AWS module:**
- Created `aws_client_id_list` local variable to avoid duplicating the client ID list in both `client_id_list` and the IAM role policy `aud` condition
- `sts.amazonaws.com` is now always included via this shared local

**Azure module:**
- Removed `oidc_issuer_uri` local - inlined the value where it's used (only one place)
- Removed `azure_blob_endpoint_suffix` local - using the literal `"core.windows.net"` directly in the output

**GCP module:**
- Fixed `GCP_PROJECT_ID` description per suggestion
- Changed `GCP_REGION` default to lowercase `europe-west4`
- Keeping variable names as-is to avoid breaking changes for existing users

**Dependabot:**
- Added entries for `/src/aws`, `/src/azure`, and `/src/gcp` directories

Regarding the suggestion to pass `random_prefix` as a variable - I've opted to keep the `random_string` resource within each module for now to keep them self-contained. Happy to revisit if there's a strong preference otherwise.

philip-gai committed:
Merge main and update provider versions

- Resolve merge conflicts
- Update AWS provider to 6.19.0
- Update Azure provider to 4.51.0, AzureAD to 3.6.0
- Update GCP provider to 7.9.0
- Update random provider to 3.7.2 across all modules

philip-gai commented on src/gcp/variables.tf:
```diff
@@ -0,0 +1,20 @@
+variable "GHES_NAME" {
```
Agreed on avoiding breaking changes. The all-caps variable names are from the original codebase, so I'm keeping them as-is to maintain backward compatibility for existing users.

philip-gai reviewed (commented):
Addressed all feedback - see individual replies on each thread.

philip-gai commented on src/gcp/variables.tf:
```diff
@@ -0,0 +1,20 @@
+variable "GHES_NAME" {
```
Agreed on avoiding breaking changes. The all-caps variable names are from the original codebase, so I'm keeping them as-is to maintain backward compatibility for existing users.

philip-gai commented on src/gcp/variables.tf:
```diff
@@ -0,0 +1,20 @@
+
+variable "GCP_PROJECT_ID" {
+  type        = string
+  description = "Google Cloud: ID of the Project to use"
```
Applied! Updated description to `"ID of the Google Cloud project to be used."`
